### PR TITLE
[WIP] Add glob deps

### LINKED ISSUE #4217: Only build needed executables in a package
Once https://github.com/haskell/cabal/pull/4055 is merged, `build-tool-depends` will be implemented, but somewhat shoddily in that on the cabal-install side the exact needed exe dependency will be thrown away and all built instead. This is extra unneeded work, and also changes the PATH in potentially unexpected ways.

Additionally, #3708 in the original issue proposed there also be wildcard deps, for depending on all exes in the package as the implementation currently does, which #4055 did not implement. Parsing and handling those would be a good thing to do in conjunction with this, lest the wildcard case in solving and building be deleted and then reimplemented.

### PR DESCRIPTION
**Pull request by Ericson2314**
This is really really WIP, but I wanted to ask whether my top commit, adding an extra field to exe deps in the solver, will help route that information to ProjectPlanning for the plan in  https://github.com/haskell/cabal/pull/4104#issuecomment-267484861, or whether I'm barking up the wrong tree.

#4217

### CODE REVIEW

**Comment by grayjay** [comment]
There is a related issue for refactoring the dependency solver to solve for components instead of packages: #4087

**Comment by Ericson2314** [comment]
@grayjay thanks! Yeah, I view this `role` as a bridge towards that---metadata ferried along by the solver for now, but actually inspected by it in the future.

**Comment by 23Skidoo** [comment]
CI failure is due to `-Werror`:

```
Distribution/Simple/Build.hs:33:1: Warning:
    The import of ‘Distribution.Types.ExeDependency’ is redundant
      except perhaps to import instances from ‘Distribution.Types.ExeDependency’
    To import instances alone, use: import Distribution.Types.ExeDependency()
```

**Comment by Ericson2314** [comment]
@23Skidoo thanks! Though heh isn't quite ready to be merged anyways :)

**Comment by ezyang** [comment]
The top commit seems fine to me. What happens if I depend on multiple executables from the same package; does it show up twice?

**Comment by Ericson2314** [comment]
@ezyang hmm probably? I haven't looked at the solving code in much detail---I mainly just changed the definition of `Dep` and then dealt with the fallout as ghc fed it to me. Perhaps it should take a list of exes and have Monoid instance I can use somewhere?

**Comment by ezyang** [comment]
@Ericson2314 My suggestion is to take a look at the output you get after the change. (Hard to get this output? Add some logging!)

Having thought about this more, it may not be necessary to propagate executable dependencies from the solver; ProjectPlanning can recompute them from scratch. Take a look at how we handle exe_deps0 in elaborateSolverToComponents: the first thing we do run elaborateExeSolverId/elaborateExePath on them, which essentially looks up the package that defined the executable(s) and then adds them as deps. With the information you have from the package description, it would be a simple matter to filter this list down to the true dependencies.

(BTW, I think this code is probably in need of some refactoring!)

**Comment by Ericson2314** [comment]
I added this to the 2.0 milestone, because without it build-tool-depends isn't so well implemented.

**Comment by Ericson2314** [comment]
Ok, did things just in ProjectPlanning as @ezyang described (I still have my old commit in another branch for future work :)).

All that's left here is https://github.com/haskell/cabal/pull/4241/files#diff-a6de9cdd16372d3e6c4ec16d4d1703a1R575, I need to route the expanded wildcard back to cabal somehow. Does this mean adding flags to Cabal?

File: Cabal/Distribution/Simple/Configure.hs [hunk 1]
```diff
@@ -570,19 +570,31 @@ configure (pkg_descr0', pbi) cfg = do
...
+          -- 2. `build-tool-depends` that aren't from the current package.
+          let externBuildToolDeps =
+                [ LegacyExeDependency (unUnqualComponentName eName) versionRange
+                | buildTool@(ExeDependency _ (Just eName) versionRange)
+                  <- getAllToolDependencies pkg_descr bi
+                , not $ isInternal pkg_descr buildTool ]
+          -- Second, we collect any build-tools entry we don't know how to
+          -- desugar. We'll never have no idea how to build them, so we just
+          -- hope they are already on the PATH.
+          let unknownBuildTools =
```
Comment (ezyang) [inline Cabal/Distribution/Simple/Configure.hs]: I just realized that I may have communicated the situation here a little poorly. It is true that Setup was modified to pick up unknown build-tools from PATH, but in fact, all the testing code we have does not seem to actually exercise this case, as it is always assumed the build-tools refers to an internal dep. So a test here would be appreciated.

Comment (Ericson2314) [inline Cabal/Distribution/Simple/Configure.hs]: Will do.

**Comment by Ericson2314** [comment]
@ezyang Yes, definitely tests, but I do need the route the expanded wildcard back to Cabal too, right?

**Comment by ezyang** [comment]
I guess so (although it's not altogether clear what that looks like; I don't think Cabal has any way of conveniently finding out what the wildcard will glob).

**Comment by Ericson2314** [comment]
I think the bug-fix portion of this should at least make 2.0 if not wildcards, so I'm going to factor those apart.

**Comment by ezyang** [comment]
Hello @Ericson2314, what would you like to be done on this PR specifically?

**Comment by Ericson2314** [comment]
@ezyang It's in limbo until someone has a good idea for routing the glob expansion back to Cabal. Not a priority for 2.0 I'd think unless @hvr differs (it's back-compat with explicit `:*` syntax).

**Comment by ezyang** [comment]
OK. Maybe this presents the case for extending the database to also include executables, or maybe teaching Cabal how to interact with Nix in some way.

**Comment by gbaz** [comment]
#4217 looks like it was resolved without this pr. can this be closed?

**Comment by jneira** [comment]
closing optimistically after no feedback for the previous comment, thanks all!

**Pull request closed without merging**